core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala

```diff
@@ -1,24 +1,104 @@
 
+import scala.collection.JavaConverters._
 import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+class DataRepoEntityProvider(requestArguments: EntityRequestArguments, workspaceManagerDAO: WorkspaceManagerDAO, dataRepoDAO: DataRepoDAO)
+  extends EntityProvider with LazyLogging {
 
-class DataRepoEntityProvider extends EntityProvider {
+  val workspace = requestArguments.workspace
+  val userInfo = requestArguments.userInfo
+  val dataReferenceName = requestArguments.dataReference.getOrElse(throw new DataEntityException("data reference must be defined for this provider"))
 
   override def entityTypeMetadata(): Future[Map[String, EntityTypeMetadata]] = {
-    // TODO: contact WSM to retrieve the data reference specified in the request
-    // TODO: extract the TDR snapshot ID from the WSM response
-    // TODO: contact TDR to describe the snapshot
-    // TODO: reformat TDR's response into the expected response structure
-    throw new UnsupportedEntityOperationException("type metadata will be supported by this provider, but is not implemented yet")
+
+    // TODO: AS-321 auto-switch to see if the ref supplied in argument is a UUID or a name?? Use separate query params? Never allow ID?
+
+    // get snapshotId from reference name
+    val snapshotId = lookupSnapshotForName(dataReferenceName)
+
+    // contact TDR to describe the snapshot
+    val snapshotModel = dataRepoDAO.getSnapshot(snapshotId, userInfo.accessToken)
+
+    // reformat TDR's response into the expected response structure
+    val entityTypesResponse: Map[String, EntityTypeMetadata] = snapshotModel.getTables.asScala.map { table =>
+      val attrs: Seq[String] = table.getColumns.asScala.map(_.getName)
+      // Is there a better way to describe the primary key if null or more than one PK, if TDR ever supports this.
+      // currently, TDR returns null for PKs in snapshots, and this is expected according to the TDR team,
+      // with no concrete plans to change.
+      val primaryKey = Option(table.getPrimaryKey).getOrElse(new java.util.ArrayList()).asScala.mkString(",")
+      // TODO: AS-321 once DR-1003 is implemented, add the actual row counts
+      (table.getName, EntityTypeMetadata(-1, primaryKey, attrs))
+    }.toMap
+
+    Future.successful(entityTypesResponse)
+
   }
 
   override def createEntity(entity: Entity): Future[Entity] =
     throw new UnsupportedEntityOperationException("create entity not supported by this provider.")
 
   override def deleteEntities(entityRefs: Seq[AttributeEntityReference]): Future[Int] =
     throw new UnsupportedEntityOperationException("delete entities not supported by this provider.")
+
+  // not marked as private to ease unit testing
+  def lookupSnapshotForName(dataReferenceName: String): UUID = {
+    // contact WSM to retrieve the data reference specified in the request
+    val dataRef = workspaceManagerDAO.getDataReferenceByName(UUID.fromString(workspace.workspaceId),
```

Comment:
I believe by the time we get here, we've asked Sam whether the user has Read access to the workspace. Since snapshots have their own sam permissions - do we need to check that the user also has permissions on the snapshot? Or are we assuming that when a snapshot is associated with a workspace, that means anyone who can or will in the future have Read access to the workspace, they should be able to see the snapshot?

Reply:
@ansingh7115 per PO business requirements, any user who has Read access to the workspace should be able to see all snapshot references in that workspace. This doesn't necessarily mean that the user can see the snapshot *contents* - and those permission checks will happen as soon as we connect to Data Repo. But as far as this specific `lookupSnapshotForName` method - any workspace user should be able to resolve the name into the snapshot UUID.